rasterio/_io.pyx

```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
         return sample_gen(self, xy, indexes)
 
 
+cdef class MemoryFileBase(object):
+    """Base for a BytesIO-like class backed by an in-memory file."""
+
+    def __init__(self, initial_bytes=b''):
+        """Map bytes to a file in an in-memory filesystem."""
+        cdef const char *path = NULL
+        cdef VSILFILE *vsi_handle = NULL
+
+        if not isinstance(initial_bytes, (bytearray, bytes)):
+            raise ValueError("Initial bytes must be type bytearray or bytes")
+
+        self.name = os.path.join('/vsimem', str(uuid.uuid4()))
+        self._pos = 0
+
+        self._initial_bytes = initial_bytes
+        cdef unsigned char *buffer = self._initial_bytes
+
+        if self._initial_bytes:
+            path_b = self.name.encode('utf-8')
+            path = path_b
+            vsi_handle = VSIFileFromMemBuffer(path, buffer,
+                                              len(self._initial_bytes), 0)
+            if vsi_handle == NULL:
+                raise OSError('failed to map buffer to file')
+            if VSIFCloseL(vsi_handle) != 0:
+                raise OSError('failed to close mapped file handle')
+
+    def check(self):
+        """True if the in-memory file exists"""
+        path_b = self.name.encode('utf-8')
+        cdef const char *path = path_b
+
+        cdef VSILFILE *fp = VSIFOpenL(path, 'r')
+        if fp == NULL:
+            result = False
+        else:
+            VSIFCloseL(fp)
+            result = True
+
+        return result
+
+    def __len__(self):
+        return self.getbuffer().size
+
+    def close(self):
+        """Close MemoryFile and unlink its in-memory file."""
```

Comment:
'unlink' doesn't seem immediately obvious, but perhaps makes sense in file system context.  Is the effect that the memory is released (eventually)?  Maybe 'Close MemoryFile and release memory' would be more intuitive.